daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -485,47 +487,48 @@ object ScenarioLedger {
                           val gk = GlobalKey.assertBuild(nc.templateId, keyWithMaintainers.key)
                           newCache1.activeKeys.get(gk) match {
                             case None => Right(newCache1.addKey(gk, nc.coid))
-                            case Some(_) => Left(UniqueKeyViolation(gk))
+                            case Some(_) => Left(CommitError.UniqueKeyViolation(gk))
                           }
                       }
                       processNodes(mbNewCache2, idsToProcess)
 
-                    case Node.Fetch(referencedCoid, templateId @ _, _, _, _, _, _, _) =>
+                    case Node.Fetch(referencedCoid, templateId @ _, _, _, _, _, _, _) if newCache.activeContracts.contains(referencedCoid) =>
```

Comment:
I think we should take @carlpulley-da’s work into account here and not merge this into the loop. Instead call `richTr.inputContracts` to get the set of contracts and check activeness of that set outside of this loop.